payjoin/src/core/uri/url_ext.rs

```diff
@@ -109,68 +112,139 @@ pub fn parse_with_fragment(endpoint: &str) -> Result<Url, BadEndpointError> {
     Ok(url)
 }
 
-fn get_param<F, T>(url: &Url, prefix: &str, parse: F) -> Option<T>
-where
-    F: Fn(&str) -> Option<T>,
-{
-    if let Some(fragment) = url.fragment() {
-        for param in fragment.split('+') {
-            if param.starts_with(prefix) {
-                return parse(param);
-            }
+#[derive(Debug)]
+pub(crate) enum ParseFragmentError {
+    InvalidChar(char),
+    AmbiguousDelimiter,
+}
+
+impl std::error::Error for ParseFragmentError {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> { None }
+}
+
+impl std::fmt::Display for ParseFragmentError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        use ParseFragmentError::*;
+
+        match &self {
+            InvalidChar(c) => write!(f, "invalid character: {c} (must be uppercase)"),
+            AmbiguousDelimiter => write!(f, "ambiguous fragment delimiter (both + and - found)"),
         }
     }
-    None
 }
 
-fn set_param(url: &mut Url, prefix: &str, param: &str) {
-    let fragment = url.fragment().unwrap_or("");
-    let mut fragment = fragment.to_string();
-    if let Some(start) = fragment.find(prefix) {
-        let end = fragment[start..].find('+').map_or(fragment.len(), |i| start + i);
-        fragment.replace_range(start..end, "");
-        if fragment.ends_with('+') {
-            fragment.pop();
+fn check_fragment_delimiter(fragment: &str) -> Result<char, ParseFragmentError> {
+    // For backwards compatibility, also accept `+` as a
+    // fragment parameter delimiter. This was previously
+    // specified, but may be interpreted as ` ` by some
+    // URI parsoing libraries. Therefore if `-` is missing,
+    // assume the URI was generated following the older
+    // version of the spec.
+
+    let has_dash = fragment.contains('-');
+    let has_plus = fragment.contains('+');
+
+    // Even though fragment is a &str, it should be ascii so bytes() correspond
+    // to chars(), except that it's easier to check that they are in range
+    for c in fragment.bytes() {
+        // These character ranges are more permissive than uppercase bech32, but
+        // also more restrictive than bech32 in general since lowercase is not
+        // allowed
+        if !(b'0'..b'9' + 1).contains(&c)
+            && !(b'A'..b'Z' + 1).contains(&c)
+            && c != b'-'
+            && c != b'+'
+        {
+            return Err(ParseFragmentError::InvalidChar(c.into()));
         }
```

Comment:
What do you think of this syntax?

```suggestion
        if !matches!(c, b'0'..=b'9' | b'A'..=b'Z' | b'-' | b'+') {
            return Err(ParseFragmentError::InvalidChar(c.into()));
        }
```